generate_workspace/src/main/java/com/google/devtools/build/workspace/maven/VersionResolver.java

```diff
@@ -123,4 +126,39 @@ private static String makeVersionRange(String versionSpec) {
   public static VersionResolver defaultResolver() {
     return new VersionResolver(Aether.defaultOption());
   }
+
+  /**
+   * Takes a version specification (as defined in
+   * http://maven.apache.org/enforcer/enforcer-rules/versionRanges.html) and finds a valid version
+   * that is likely to exist.  Basically: 1.2.3 is 1.2.3+, [1.2.3] is exactly 1.2.3, and then
+   * there is comma-separated range notation.
+   */
+  @Nullable
+  static String resolveVersionLocally(String unparsedVersion) {
+    VersionRange versionRange;
+    try {
+      versionRange = VersionRange.createFromVersionSpec(unparsedVersion);
+    } catch (InvalidVersionSpecificationException e) {
+      return null;
+    }
+    if (versionRange.getRecommendedVersion() != null) {
+      return versionRange.getRecommendedVersion().toString();
+    }
```

Comment:
**Feedback:** Can you separate out the above lines to a separate function, and remove the org.apache.maven.artifact.versioning classes?

When I was initially reading the code a month or two ago, this portion was very confusing to me. Especially the `getRecommendedVersion` portion, and the versionRange restrictions in 149. 

**Proposed logic for function**: I believe the logic should be 
```java
static String resolveVersionLocally(String versionSpecification) {
    if ( isPin(versionSpecification) ) {
         return getPinnedVersion(versionSpecification);
    }
    return null;
}
```
The comment below argues why I believe we shouldn't locally resolve non-pinned version specs.

**Why remove versioning package**: Supposing you agree to that, it doesn't make sense to introduce the org.apache.maven.artifact.versioning package.

Identifying a pin (hard or soft) is a simple enough task. However, by including this package we add unnecessary mental overhead. For example, why are there multiple restrictions of a VersionRange? what does it mean to be a "restriction"? What is "recommended version" by maven? When is the recommended version invalid? Why is the recommended version invalid for soft pins but not for hard pins or version ranges? 

Previously, I had to dig through the source code and I left with a meh impression of the code quality. All we really need in this context is

```java
private boolean isPin(String versionSpec) {
     return !isVersionRange(versionSpec) || isHardPin(versionSpec);
}
private boolean isHardPin(String versionSpec) {
    if (versionSpec == null) return false;
    int length = versionSpec.length();
    return versionSpec.charAt(0) == '[' && versionSpec.charAt(length -1) == ']'
}
```

Sure we would have to write a few tests. But at least we don't have to inherit maven.versionings overheard and idiosyncratic development schemes.